src/sentry/tasks/unmerge.py

```diff
@@ -0,0 +1,468 @@
+    for event in events:
+        counters[event.datetime][tsdb.models.group][event.group_id] += 1
+
+        user = event.data.get('sentry.interfaces.User')
+        if user:
+            sets[event.datetime][tsdb.models.users_affected_by_group][event.group_id].add(
+                EventUser(
+                    project=project,
+                    ident=user.get('id'),
+                    email=user.get('email'),
+                    username=user.get('username'),
+                    ip_address=user.get('ip_address'),
+                ).tag_value
+            )
+
+        environment = Environment.objects.get(
+            projects=project,
+            name=event.data.get('environment', ''),
+        )
+
+        frequencies[event.datetime][tsdb.models.frequent_environments_by_group][event.group_id][environment.id] += 1
+
+        # TODO: Double check this!
+        release = event.get_tag('sentry:release')
+        if release:
+            # TODO: I'm also not sure if this is correct, see similar comment
+            # above during creation.
+            environment = event.data.get('environment', '')
+
+            # XXX: This is also inefficient, especially since we have created
+            # or updated the record already in this process.
+            grouprelease = GroupRelease.objects.get(
+                group_id=event.group_id,
+                environment=environment,
+                release_id=Release.objects.get(
+                    organization_id=project.organization_id,
+                    version=release,
+                ).id,
+            )
+
+            frequencies[event.datetime][tsdb.models.frequent_releases_by_group][event.group_id][grouprelease.id] += 1
+
+    return counters, sets, frequencies
+
+
+def repair_tsdb_data(project, events):
+    counters, sets, frequencies = collect_tsdb_data(project, events)
+
+    for timestamp, data in counters.items():
+        for model, keys in data.items():
+            for key, value in keys.items():
+                tsdb.incr(model, key, timestamp, value)
+
+    for timestamp, data in sets.items():
+        for model, keys in data.items():
+            for key, values in keys.items():
+                # TODO: This should use `record_multi` rather than `record`.
+                tsdb.record(model, key, values, timestamp)
+
+    for timestamp, data in frequencies.items():
+        tsdb.record_frequency_multi(data.items(), timestamp)
+
+
+def repair_denormalizations(project, events):
+    repair_tag_data(project, events)
+    repair_group_release_data(project, events)
+    repair_tsdb_data(project, events)
+
+
+def update_tag_value_counts(id_list):
+    instances = GroupTagKey.objects.filter(group_id__in=id_list)
+    for instance in instances:
+        instance.update(
+            values_seen=GroupTagValue.objects.filter(
+                project_id=instance.project_id,
+                group_id=instance.group_id,
+                key=instance.key,
+            ).count(),
```

Comment:
> I sorta wondered if we could instead just update this as we go in repair_tag_data, and we could, but it would mean a lot more writes in high-tag-value-cardinality cases, so almost definitely not worth it.

That was my first thought too, but that introduces two issues:

1. tracking high cardinality values (as you mentioned) could become expensive,
2. it also increases the potential for inconsistency here due to data races because any new tag keys that were associated with this issue after the `GroupHash` was moved will be ignored.

>  How hard is it to get this to run as a subquery? I guess that's probably a django ORM question?

Not sure, but also a little hesitant to add the complexity here of dropping to raw SQL just for the sake of optimizing this into a single transaction. (I might take a look though in case I'm wrong.)